esphome/components/remote_base/midea_protocol.h

```diff
@@ -0,0 +1,203 @@
+  // Set power
+  void set_power(bool val) { set_value_(1, 0b1, 7, val); }
+  // Set ON timer (0: disable)
+  void set_on_timer(uint16_t minutes);
+  // Set OFF timer (0: disable)
+  void set_off_timer(uint16_t minutes);
+
+ protected:
+  static const uint8_t MIN_TEMP = 17;
+  static const uint8_t MAX_TEMP = 30;
+};
+
+class MideaProtocol : public RemoteProtocol<MideaData> {
+ public:
+  void encode(RemoteTransmitData *dst, const MideaData &data) override;
+  optional<MideaData> decode(RemoteReceiveData src) override;
+  void dump(const MideaData &data) override;
+
+ protected:
+  static const int32_t TICK_US = 560;
+  static const int32_t HEADER_HIGH_US = 8 * TICK_US;
+  static const int32_t HEADER_LOW_US = 8 * TICK_US;
+  static const int32_t BIT_HIGH_US = 1 * TICK_US;
+  static const int32_t BIT_ONE_LOW_US = 3 * TICK_US;
+  static const int32_t BIT_ZERO_LOW_US = 1 * TICK_US;
+  static const int32_t MIN_GAP_US = 10 * TICK_US;
+  static void one(RemoteTransmitData *dst) { dst->item(BIT_HIGH_US, BIT_ONE_LOW_US); }
+  static void zero(RemoteTransmitData *dst) { dst->item(BIT_HIGH_US, BIT_ZERO_LOW_US); }
+  static void header(RemoteTransmitData *dst) { dst->item(HEADER_HIGH_US, HEADER_LOW_US); }
+  static void footer(RemoteTransmitData *dst) { dst->item(BIT_HIGH_US, MIN_GAP_US); }
+  static void data(RemoteTransmitData *dst, const MideaData &src, bool compliment = false);
+  static bool expect_one(RemoteReceiveData &src);
+  static bool expect_zero(RemoteReceiveData &src);
+  static bool expect_header(RemoteReceiveData &src);
+  static bool expect_footer(RemoteReceiveData &src);
+  static bool expect_data(RemoteReceiveData &src, MideaData &out);
+};
+
+DECLARE_REMOTE_PROTOCOL(Midea)
+
+template<typename... Ts> class MideaRawAction : public RemoteTransmitterActionBase<Ts...> {
+ public:
+  void set_data(const uint8_t *data) {
+    this->data_ = data;
+    this->data_.finalize();
+  }
+  void set_template(std::function<std::vector<uint8_t>(Ts...)> func) { this->code_func_ = func; }
+  void encode(RemoteTransmitData *dst, Ts... x) override {
+    if (this->code_func_ != nullptr) {
+      this->data_ = this->code_func_(x...);
+      this->data_.finalize();
+    }
+    MideaProtocol().encode(dst, this->data_);
+  }
+
+ protected:
+  std::function<std::vector<uint8_t>(Ts...)> code_func_{};
+  MideaData data_;
+};
+
+template<typename... Ts> class MideaFollowMeAction : public RemoteTransmitterActionBase<Ts...> {
+ public:
+  void set_template(std::function<uint8_t(Ts...)> func) { this->code_func_ = func; }
+  void set_beeper(bool value) { this->data_.set_beeper(value); }
+  void set_temp(uint8_t temp) {
+    data_.set_temp(temp);
+    data_.finalize();
+  }
+  void encode(RemoteTransmitData *dst, Ts... x) override {
+    if (this->code_func_ != nullptr)
+      set_temp(this->code_func_(x...));
+    MideaProtocol().encode(dst, this->data_);
+  }
+
+ protected:
+  std::function<uint8_t(Ts...)> code_func_{};
+  MideaFollowMe data_;
+};
```

Comment:
This action can be simplified like this:

```suggestion
  TEMPLATABLE_VALUE(uint8_t, temperature)
  TEMPLATABLE_VALUE(boolean, beeper)

  }
  void encode(RemoteTransmitData *dst, Ts... x) override {
    MideaFollowMe data_;
    data_.set_temp(this->temperature_.value(x...);
    data_.set_beeper(this->beeper_.value(x...);
    data_.finalize();
    MideaProtocol().encode(dst, data_);
  }
};
```